### ISSUE TITLE
Can you explain the loss function used for training ? what are the steps you followed for data pre-processing and what does the shape and the label look like.

### ISSUE THREAD

**Issue by anand-kummari** [fabvio/Cascade-LD#4]

**Comment by fabvio** [comment]
Hi,
1) I shared the loss function in the latest commit. It is an implementation of Hsu clustering loss as explained in the paper. Basically, it minimizes the KL divergence between points belonging to the same cluster.
2) For preprocessing, I downsample the images to 512x256 and apply random horizontal flips.
3) The label is a projection of the TuSimple labels on images for the segmentation task, and simple class IDs for classification. For segmentation, I simply draw the polylines on a blank image with a width of 2 pixels per side. Each lane (maximum 4 lanes for image) has an unique ID. The class IDs are associated to the detected labels taking the minimum average distance one with respect to the ground truth, if existing. With this strategy, it is possible to extract the descriptors with the code in the notebook.

**Comment by anand-kummari** [comment]
can you share a step by step procedure for generating labels for segmentation and classification. I have generated some labels and the output after training for 150 epochs is not correct. I wanted an ouput with each lane in one class but I get some part of every lane in one class. So this is not a correct output.
currently I draw polylines on blank 3 channeled image with each line in different color with a width of 5 and use them as labels.
Please share your approach to create labels which can be used to find loss using the loss function you provided.

**Comment by fabvio** [comment]
I can assist you in getting a good result. First of all, your result seems to be dependent also on the training. Could you share an example of your labels and output? Then, are you training from scratch or you are fine tuning a previously trained model on the binary segmentation?

**Comment by fabvio** [comment]
Also, what you want are labels on a greyscale image with contiguous ids, as you do in segmentation. Why do you use three channels?

**Comment by anand-kummari** [comment]
![0313-1_360](https://user-images.githubusercontent.com/26199198/65936945-eeac8d80-e43b-11e9-996b-47fde66f78f1.jpg)

This is one of my label image obtained from Tusimple data,used during training. Below is the output image at testing. I expected the output to be the entire lane in one class. But the output I get has all lanes in one class shown in the image below
![6_epoch_enc_lane_out](https://user-images.githubusercontent.com/26199198/65937135-7db9a580-e43c-11e9-9bcd-470e68e7d619.jpg)
![result](https://user-images.githubusercontent.com/26199198/65938050-81026080-e43f-11e9-8d2e-0acc1d9978cc.jpg)

I am training the model from scratch, training on tusimple dataset, using the loss function that you provided. Please let me know if I am doing wrong anywhere. Also is my label image good for calculating loss function given by you? If not please specify how the label image should be.

**Comment by fabvio** [comment]
Well, if the label is RGB you are already doing something wrong, as I told you in my previous message. Your label must be greyscale, with class IDs going from 1 to 5 with 0 as background. Are you showing some colourized version of your label? Also, it is *fundamental* that you first pretrain your model with a standard cross entropy loss on binary segmentation (background/lane boundary). The result you get appears highly related to that. The KL loss seems to be highly sensitive for this task, so training from scratch won't lead to working results. Please let me know how it goes.

**Comment by anand-kummari** [comment]
Can you please tell me the epochs, and the parameters for pretraining the model with standard cross entropy loss like learning rate, momentum etc.

**Comment by fabvio** [comment]
It's a standard erfnet training. If I recall correctly from the paper, 150 epochs, optimized Adam with learning rate 5e-4, default b1 and b2.

**Comment by fabvio** [comment]
Closing as I suppose the problem is solved.
